File: android/app/src/main/java/com/propertyfinder1/FlexViewLayouter.java
package com.propertyfinder1;

import com.propertyfinder1.util.IntervalTree;
import com.propertyfinder1.util.Interval;

import android.view.View;
import java.util.HashMap;
import java.util.ArrayList;

public class FlexViewLayouter {
    private boolean horizontal;
    private IntervalTree intervalTree;
    private HashMap<String, Interval> intervalPositions;

    public FlexViewLayouter(boolean horizontal) {
        this.horizontal = horizontal;
        this.intervalTree = new IntervalTree();
        this.intervalPositions = new HashMap<>();
    }

    /*
    The algorithm which finds the overlaps and shifts the views accordingly. A Binary search tree is used to store the boundary of the views
    which have layouted so far. First, the left / x position of the view is corrected if overlaps are found such that xDiff > yDiff. Once the
    left position is corrected, new overlaps are checked for the updated position. If overlaps are found in the y-direction, then the view is
    shifted by an amount equal to the largest overlap. The views which go outside the bounds of the container will be clipped as the parent
    scrollview enforces the container view to have the size as specified by the user.
    */
    public void layout(View view) {
        int height = view.getHeight();
        int width = view.getWidth();
        int top = view.getTop();
        int left = view.getLeft();
        Interval currentInterval = horizontal ? new Interval(top, top + height) : new Interval(left, left + width);

        int correctedTop = top;
        int correctedLeft = left;
        ArrayList<Interval> overlappingIntervals = intervalTree.searchAll(currentInterval.getStart(), currentInterval.getEnd());
        for (int i = 0; i < overlappingIntervals.size(); i++) {
            Interval overlappingInterval = overlappingIntervals.get(i);
            Interval overlappingIntervalPosition = intervalPositions.get(overlappingInterval.toString());
            if (overlappingIntervalPosition == null) {
                continue;
            }
            if (horizontal) {
                int xDiff = Math.abs(correctedLeft - overlappingIntervalPosition.getStart());
                int yDiff = Math.abs(correctedTop - overlappingInterval.getStart());
                if (yDiff > xDiff){
                    correctedTop = Math.max(correctedTop, overlappingInterval.getEnd());
                }
            } else {
                int xDiff = Math.abs(correctedLeft - overlappingInterval.getStart());
                int yDiff = Math.abs(correctedTop - overlappingIntervalPosition.getStart());
                if (xDiff > yDiff) {
                    correctedLeft = Math.max(correctedLeft, overlappingInterval.getEnd());
                }
            }
        }
        if (horizontal) {
            currentInterval.setStart(correctedTop);
            currentInterval.setEnd(correctedTop + height);
        } else {
            currentInterval.setStart(correctedLeft);
            currentInterval.setEnd(correctedLeft + width);
        }

        overlappingIntervals = intervalTree.searchAll(currentInterval.getStart(), currentInterval.getEnd());
        for (int i = 0; i < overlappingIntervals.size(); i++) {
            Interval overlappingInterval = overlappingIntervals.get(i);
            Interval overlappingIntervalPosition = intervalPositions.get(overlappingInterval.toString());
            if (overlappingIntervalPosition == null) {
                continue;
            }
            if (horizontal) {
                correctedLeft = Math.max(correctedLeft, overlappingIntervalPosition.getEnd());
            } else {
                correctedTop = Math.max(correctedTop, overlappingIntervalPosition.getEnd());
            }

            if (currentInterval.equals(overlappingInterval)) {
                if (horizontal) {
                    updateInterval(currentInterval, new Interval(correctedLeft, correctedLeft + width));
                } else {
                    updateInterval(currentInterval, new Interval(correctedTop, correctedTop + height));
                }
                view.layout(correctedLeft, correctedTop, correctedLeft + width, correctedTop + height);
                return;
            }
            if (currentInterval.contains(overlappingInterval)) {
                removeInterval(overlappingInterval);
            } else if (overlappingInterval.contains(currentInterval)) {
                removeInterval(overlappingInterval);
                addInterval(new Interval(overlappingInterval.getStart(),currentInterval.getStart()), overlappingIntervalPosition);
                addInterval(new Interval(currentInterval.getEnd(),overlappingInterval.getEnd()), overlappingIntervalPosition);
            } else if (overlappingInterval.compareTo(currentInterval) < 0){
                removeInterval(overlappingInterval);
                addInterval(new Interval(overlappingInterval.getStart(),currentInterval.getStart()), overlappingIntervalPosition);
            } else {
                removeInterval(overlappingInterval);
                addInterval(new Interval(currentInterval.getEnd(), overlappingInterval.getEnd()), overlappingIntervalPosition);
            }
        }
        if (horizontal) {
            addInterval(currentInterval, new Interval(correctedLeft, correctedLeft + width));
        } else {
            addInterval(currentInterval, new Interval(correctedTop, correctedTop + height));
        }
        view.layout(correctedLeft, correctedTop, correctedLeft + width, correctedTop + height);
    }

    private void addInterval(Interval interval, Interval intervalPosition) {
        if(interval.getStart() == interval.getEnd()) {
            return;
        }
        intervalTree.add(interval);
        intervalPositions.put(interval.toString(), intervalPosition);
    }

    private void removeInterval(Interval interval) {
        intervalTree.remove(interval);
        intervalPositions.remove(interval.toString());
    }

    private void updateInterval(Interval interval, Interval intervalPosition) {
        intervalPositions.put(interval.toString(), intervalPosition);
    }

}

File: ios/FlexViewLayouter.swift
//
//  FlexViewLayouter.swift
//  PropertyFinder1
//
//  Created by K Arun Kumar Reddy on 19/12/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation

class FlexViewLayouter {
    private var isHorizontal: Bool;
    private var intervalTree: IntervalTree;
    private var intervalPositions: [String: Interval];
    
    init(isHorizontal: Bool) {
        self.isHorizontal = isHorizontal;
        self.intervalTree = IntervalTree();
        self.intervalPositions = [String: Interval]();
    }
    
    func layout(view: UIView) {
        let height: Int = Int(view.frame.height);
        let width: Int = Int(view.frame.width);
        let left: Int = Int(view.frame.minX);
        let top: Int = Int(view.frame.minY);
        let currentInterval = isHorizontal ? Interval(start: top, end: top + height) : Interval(start: left, end: left + width);
        
        var correctedTop: Int = top;
        var correctedLeft: Int = left;
        var overlappingIntervals: Array<Interval> = intervalTree.searchAll(start: currentInterval.getStart(), end: currentInterval.getEnd());
        for overlappingInterval in overlappingIntervals {
            if let overlappingIntervalPosition: Interval = intervalPositions[overlappingInterval.description] {
                if (isHorizontal) {
                    let xDiff: Int = abs(correctedLeft - overlappingIntervalPosition.getStart());
                    let yDiff: Int = abs(correctedTop - overlappingInterval.getStart());
                    if (yDiff > xDiff){
                        correctedTop = max(correctedTop, overlappingInterval.getEnd());
                    }
                } else {
                    let xDiff: Int = abs(correctedLeft - overlappingInterval.getStart());
                    let yDiff: Int = abs(correctedTop - overlappingIntervalPosition.getStart());
                    if (xDiff > yDiff) {
                        correctedLeft = max(correctedLeft, overlappingInterval.getEnd());
                    }
                }
                
            }
        }
        if (isHorizontal) {
            currentInterval.setStart(start: correctedTop);
            currentInterval.setEnd(end: correctedTop + height);
        } else {
            currentInterval.setStart(start: correctedLeft);
            currentInterval.setEnd(end: correctedLeft + width);
        }
        overlappingIntervals = intervalTree.searchAll(start: currentInterval.getStart(), end: currentInterval.getEnd());
        for overlappingInterval in overlappingIntervals {
            if let overlappingIntervalPosition: Interval = intervalPositions[overlappingInterval.description] {
                if (isHorizontal) {
                    correctedLeft = max(correctedLeft, overlappingIntervalPosition.getEnd());
                } else {
                    correctedTop = max(correctedTop, overlappingIntervalPosition.getEnd());
                }
                
                if (currentInterval == overlappingInterval) {
                    if (isHorizontal) {
                        updateInterval(interval: currentInterval, intervalPosition: Interval(start: correctedLeft, end: correctedLeft + width));
                    } else {
                        updateInterval(interval: currentInterval, intervalPosition: Interval(start: correctedTop, end: correctedTop + height));
                    }
                    view.frame = CGRect(x: correctedLeft, y:correctedTop, width: width, height: height);
                    return;
                }
                if (currentInterval.contains(interval: overlappingInterval)) {
                    removeInterval(interval: overlappingInterval);
                } else if (overlappingInterval.contains(interval: currentInterval)) {
                    removeInterval(interval: overlappingInterval);
                    addInterval(interval: Interval(start: overlappingInterval.getStart(), end: currentInterval.getStart()), intervalPosition: overlappingIntervalPosition);
                    addInterval(interval: Interval(start: currentInterval.getEnd(),end: overlappingInterval.getEnd()), intervalPosition: overlappingIntervalPosition);
                } else if (overlappingInterval.compareTo(interval: currentInterval) < 0) {
                    removeInterval(interval: overlappingInterval);
                    addInterval(interval: Interval(start: overlappingInterval.getStart(), end: currentInterval.getStart()), intervalPosition: overlappingIntervalPosition);
                } else {
                    removeInterval(interval: overlappingInterval);
                    addInterval(interval: Interval(start: currentInterval.getEnd(), end: overlappingInterval.getEnd()), intervalPosition: overlappingIntervalPosition);
                }
            }
        }
        if (isHorizontal) {
            addInterval(interval: currentInterval, intervalPosition: Interval(start: correctedLeft, end: correctedLeft + width));
        } else {
            addInterval(interval: currentInterval, intervalPosition: Interval(start: correctedTop, end: correctedTop + height));
        }
        view.frame = CGRect(x: correctedLeft, y:correctedTop, width: width, height: height);
    }
    
    private func addInterval(interval: Interval, intervalPosition: Interval) {
        if(interval.getStart() == interval.getEnd()) {
            return;
        }
        intervalTree.add(interval: interval);
        intervalPositions[interval.description] = intervalPosition;
    }
    
    private func removeInterval(interval: Interval) {
        intervalTree.add(interval: interval);
        intervalPositions.removeValue(forKey: interval.description);
    }
    
    private func updateInterval(interval: Interval, intervalPosition: Interval) {
        intervalPositions[interval.description] = intervalPosition;
    }
}
